src/core/machine-id-setup.c

```diff
@@ -142,13 +143,35 @@ int machine_id_setup(const char *root, sd_id128_t machine_id, sd_id128_t *ret) {
                 if (ftruncate(fd, 0) < 0)
                         return log_error_errno(errno, "Failed to truncate %s: %m", etc_machine_id);
 
-                if (id128_write_fd(fd, ID128_PLAIN, machine_id, true) >= 0)
-                        goto finish;
+                /* If the caller requested a transient machine-id, write the string "uninitialized\n" to
+                 * disk and overmount it with a transient file.
+                 *
+                 * Otherwise write the machine-id directly to disk. */
+                if (force_transient) {
+                        r = loop_write(fd, "uninitialized\n", strlen("uninitialized\n"), false);
+                        if (r < 0) {
+                                return log_error_errno(r, "Failed to write uninitialized %s: %m", etc_machine_id);
+                        }
+
+                        if (fsync(fd) < 0)
+                                return log_error_errno(errno, "Failed to sync %s: %m", etc_machine_id);
+
+                        r = fsync_directory_of_file(fd);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to sync directory of %s: %m", etc_machine_id);
```

Comment:
> use fsync_full() instead.

Makes sense, thanks!

> Also I think we should do it one if level up, i.e. so that we also sync things when we properly initialize the machine ID

`id128_write_fd()` already calls fsync() and fsync_directory_of_file() so it's not necessary.